### ISSUE TITLE
Timed out proxy starting the userland proxy

### ISSUE THREAD

**Issue by PortNumber53** [docker/for-mac#2156]
### Expected behavior
docker works

### Actual behavior
almost every time I had
ERROR: for e31c121514d5_coocms_cms_nginx_1  Cannot start service cms_nginx: driver failed programming external connectivity on endpoint coocms_cms_nginx_1 (881dace9daa7f18e64791fb03d41de02466314e44ffc57f9d8908912ebce330f): Timed out proxy starting the userland proxy

ERROR: for cms_nginx  Cannot start service cms_nginx: driver failed programming external connectivity on endpoint coocms_cms_nginx_1 (881dace9daa7f18e64791fb03d41de02466314e44ffc57f9d8908912ebce330f): Timed out proxy starting the userland proxy
ERROR: Encountered errors while bringing up the project.

### Information

I've tried rebooting my mac, reinstalling docker (newest), resetting all data; reinstalling; restarting things in different ways and order; removing volumes, removing container; enabling docker-sync, disabling docker-sync;  yet nothing seems to affect this consistently, so haven't been able to rule out anything...

One thing I can see is that we use Sophos and it opens a lot of connections:
SophosSXL 93704 root   13u  IPv4 0xfe60d0e2b7404a71      0t0  TCP ml-04kg8wl.redacted.local:59128->ec2-184-169-138-218.us-west-1.compute.amazonaws.com:http (ESTABLISHED)
SophosSXL 93704 root   14u  IPv4 0xfe60d0e2aea02a71      0t0  TCP ml-04kg8wl.redacted.local:59266->ec2-184-72-32-75.us-west-1.compute.amazonaws.com:http (ESTABLISHED)
SophosSXL 93704 root   15u  IPv4 0xfe60d0e2bb882369      0t0  TCP ml-04kg8wl.redacted.local:59394->ec2-184-72-32-75.us-west-1.compute.amazonaws.com:http (ESTABLISHED)
SophosSXL 93704 root   16u  IPv4 0xfe60d0e2b8fe5691      0t0  TCP ml-04kg8wl.redacted.local:59942->ec2-184-169-135-103.us-west-1.compute.amazonaws.com:http (ESTABLISHED)
SophosSXL 93704 root   17u  IPv4 0xfe60d0e2a3174a71      0t0  TCP ml-04kg8wl.redacted.local:59984->ec2-184-169-135-103.us-west-1.compute.amazonaws.com:http (ESTABLISHED)
SophosSXL 93704 root   18u  IPv4 0xfe60d0e2b9231a71      0t0  TCP ml-04kg8wl.redacted.local:59943->ec2-184-169-135-103.us-west-1.compute.amazonaws.com:http (ESTABLISHED)
SophosSXL 93704 root   19u  IPv4 0xfe60d0e2c3848179      0t0  TCP ml-04kg8wl.redacted.local:59985->ec2-184-169-135-103.us-west-1.compute.amazonaws.com:http (ESTABLISHED)
SophosSXL 93704 root   20u  IPv4 0xfe60d0e2bb87ff89      0t0  TCP ml-04kg8wl.redacted.local:59658->ec2-184-169-135-103.us-west-1.compute.amazonaws.com:http (ESTABLISHED)
SophosSXL 93704 root   21u  IPv4 0xfe60d0e2bb881179      0t0  TCP ml-04kg8wl.redacted.local:59986->ec2-184-169-135-103.us-west-1.compute.amazonaws.com:http (ESTABLISHED)
SophosSXL 93704 root   22u  IPv4 0xfe60d0e2b7b15881      0t0  TCP ml-04kg8wl.redacted.local:59987->ec2-184-169-135-103.us-west-1.compute.amazonaws.com:http (ESTABLISHED)
SophosSXL 93704 root   23u  IPv4 0xfe60d0e2ae9e1a71      0t0  TCP ml-04kg8wl.redacted.local:59659->ec2-184-169-135-103.us-west-1.compute.amazonaws.com:http (ESTABLISHED)
SophosSXL 93704 root   24u  IPv4 0xfe60d0e2b9011f89      0t0  TCP ml-04kg8wl.redacted.local:59660->ec2-184-169-135-103.us-west-1.compute.amazonaws.com:http (ESTABLISHED)
SophosSXL 93704 root   25u  IPv4 0xfe60d0e2a2004881      0t0  TCP ml-04kg8wl.redacted.local:59661->ec2-184-169-135-103.us-west-1.compute.amazonaws.com:http (ESTABLISHED)
SophosSXL 93704 root   26u  IPv4 0xfe60d0e2ac2a3c61      0t0  TCP ml-04kg8wl.redacted.local:59662->ec2-184-169-135-103.us-west-1.compute.amazonaws.com:http (ESTABLISHED)
SophosSXL 93704 root   27u  IPv4 0xfe60d0e2b6c48881      0t0  TCP ml-04kg8wl.redacted.local:59665->ec2-184-169-135-103.us-west-1.compute.amazonaws.com:http (ESTABLISHED)
SophosSXL 93704 root   28u  IPv4 0xfe60d0e2ba188179      0t0  TCP ml-04kg8wl.redacted.local:59675->ec2-184-169-135-103.us-west-1.compute.amazonaws.com:http (ESTABLISHED)
SophosSXL 93704 root   29u  IPv4 0xfe60d0e2bbec8369      0t0  TCP ml-04kg8wl.redacted.local:59988->ec2-184-169-135-103.us-west-1.compute.amazonaws.com:http (ESTABLISHED)
SophosSXL 93704 root   30u  IPv4 0xfe60d0e2a59bfc61      0t0  TCP ml-04kg8wl.redacted.local:59989->ec2-184-169-135-103.us-west-1.compute.amazonaws.com:http (ESTABLISHED)
SophosSXL 93704 root   31u  IPv4 0xfe60d0e2a56f6369      0t0  TCP ml-04kg8wl.redacted.local:59990->ec2-184-169-135-103.us-west-1.compute.amazonaws.com:http (ESTABLISHED)
SophosSXL 93704 root   32u  IPv4 0xfe60d0e2b8fe6881      0t0  TCP ml-04kg8wl.redacted.local:59991->ec2-184-169-135-103.us-west-1.compute.amazonaws.com:http (ESTABLISHED)
SophosSXL 93704 root   33u  IPv4 0xfe60d0e2a27af369      0t0  TCP ml-04kg8wl.redacted.local:59992->ec2-184-169-135-103.us-west-1.compute.amazonaws.com:http (ESTABLISHED)
SophosSXL 93704 root   34u  IPv4 0xfe60d0e2ba2f3179      0t0  TCP ml-04kg8wl.redacted.local:59993->ec2-184-169-135-103.us-west-1.compute.amazonaws.com:http (ESTABLISHED)
SophosSXL 93704 root   35u  IPv4 0xfe60d0e2a5714179      0t0  TCP ml-04kg8wl.redacted.local:59994->ec2-184-169-135-103.us-west-1.compute.amazonaws.com:http (ESTABLISHED)
SophosSXL 93704 root   36u  IPv4 0xfe60d0e2b6c2a179      0t0  TCP ml-04kg8wl.redacted.local:59995->ec2-184-169-135-103.us-west-1.compute.amazonaws.com:http (ESTABLISHED)
SophosSXL 93704 root   37u  IPv4 0xfe60d0e2b9012881      0t0  TCP ml-04kg8wl.redacted.local:59996->ec2-184-169-135-103.us-west-1.compute.amazonaws.com:http (ESTABLISHED)
SophosSXL 93704 root   38u  IPv4 0xfe60d0e2b4eedc61      0t0  TCP ml-04kg8wl.redacted.local:59997->ec2-184-169-135-103.us-west-1.compute.amazonaws.com:http (ESTABLISHED)
SophosSXL 93704 root   39u  IPv4 0xfe60d0e2ba249691      0t0  TCP ml-04kg8wl.redacted.local:60143->ec2-184-169-135-103.us-west-1.compute.amazonaws.com:http (ESTABLISHED)
SophosSXL 93704 root   40u  IPv4 0xfe60d0e2b77a6179      0t0  TCP ml-04kg8wl.redacted.local:60145->ec2-184-169-135-103.us-west-1.compute.amazonaws.com:http (ESTABLISHED)
SophosSXL 93704 root   41u  IPv4 0xfe60d0e298ea8f89      0t0  TCP ml-04kg8wl.redacted.local:60148->ec2-184-169-135-103.us-west-1.compute.amazonaws.com:http (ESTABLISHED)
SophosSXL 93704 root   42u  IPv4 0xfe60d0e2a2e14881      0t0  TCP ml-04kg8wl.redacted.local:60150->ec2-184-169-135-103.us-west-1.compute.amazonaws.com:http (ESTABLISHED)
SophosSXL 93704 root   43u  IPv4 0xfe60d0e2bb624a71      0t0  TCP ml-04kg8wl.redacted.local:60152->ec2-184-169-135-103.us-west-1.compute.amazonaws.com:http (ESTABLISHED)

### Steps to reproduce the behavior

really just trying to bring the nginx container up

**Comment by iToto** [comment]
I've also been plagued w/ this issue

**Comment by jzila** [comment]
We are also plagued by this. It makes Docker for Mac basically unusable.

**Comment by jzila** [comment]
Worse, if you `docker rm <foo>` after this crops up, it doesn't release port mappings, so you have to restart Docker for Mac to even try again.

**Comment by prasadram** [comment]
i am also facing the same issue, restart doesn't work for me

**Comment by donaldwasserman** [comment]
I've also been haunted by this (using `Version 17.09.1-ce-mac42 (21090)`).

**Comment by hsathe** [comment]
I have the same version and tried rebooting, recreating everything. None of it works for me. Anyone was able to find a solution?

**Comment by donaldwasserman** [comment]
I don't know what the problem was (although I just upgraded it), but for whatever reason I was using docker-compose .yml files to stop all my project containers. For whatever reason, occasionally 1/n services would stay up and I'd have to do `docker stop <id> && docker rm <id>`...

¯\_(ツ)_/¯

**Comment by YRM64** [comment]
A generic response to this issue, and I don't believe in generic responses, and I've read the thread going back 3 years to get a history on the problem.  Developers, Administrators, and Engineers all agree to increase the timeout periods as a workaround.  I personally don't cater to workarounds because where they work in one instance, they can cause problems elsewhere in the developmental stage of your build.  I like the comments made by @Lk4D4 3 years ago.  He and I share the same thinking on this issue ~ no workarounds, a national fix is the appropriate solution for Docker-proxy timeouts under heavy load.

**Comment by nbcsteveb** [comment]
/remove-lifecycle stale

Having the same issue here..

Our setup works fine on some machines, but others are throwing errors like this:

```
ERROR: for phpmyadmin  Cannot start service phpmyadmin: driver failed programming external connectivity on endpoint my-phpmyadmin (bf8032b286baf84c0e6a706b9a8591f42173186d7e83d61b9488285a034375c3): Timed out proxy starting the userland proxy
```

**Comment by dajer93** [comment]
I've experienced the same issue with docker for mac. The `docker-compose` commands caused `Timed out proxy starting the userland proxy` errors as described above. I resolved the problem by reinstalling docker for mac.

**Comment by lmakarov** [comment]
I was able to track this down to a very specific use case. Happens 100% of time after starting and then trying to stop a container with an **privileged port bound to an IP**, e.g.:

```
$ docker run --rm -p 127.0.0.1:80:80 httpd:alpine
...
<ctrl+c>
```

After running that, you won't be able to start any container with a published port (with or without the IP binding).

![image](https://user-images.githubusercontent.com/1205005/66691335-74a1b300-ec4a-11e9-9204-631cf71d24c6.png)

Tested on the most recent stable Docker Desktop v2.1.0.3 (38240), clean (factory reset).
